lib/Sema/TypeCheckDeclPrimary.cpp

```diff
@@ -3670,40 +3670,114 @@ void TypeChecker::checkParameterList(ParameterList *params,
+    ArrayRef<TypeRepr *> specializationArgs,
+    DeclContext *dc, constraints::ConstraintSystem &cs) {
+  using namespace constraints;
+  // Resolve each type.
+  SmallVector<Type, 2> specializationArgTypes;
+  auto options =
+      TypeResolutionOptions(TypeResolverContext::InExpression);
+  for (auto specializationArg : specializationArgs) {
+    const auto result = TypeResolution::resolveContextualType(
+        specializationArg, dc, options,
+        // Introduce type variables for unbound generics.
+        OpenUnboundGenericType(cs, locator),
+        HandlePlaceholderType(cs, locator),
+        OpenPackElementType(cs, locator, nullptr));
+    if (result->hasError())
+      return true;
+
+    specializationArgTypes.push_back(result);
+  }
+  cs.addConstraint(
+      ConstraintKind::ExplicitGenericArguments, boundType,
+      PackType::get(cs.getASTContext(), specializationArgTypes),
+      locator);
+  return false;
+}
+
+BraceStmt *
 ExpandMacroExpansionDeclRequest::evaluate(Evaluator &evaluator,
                                           MacroExpansionDecl *MED) const {
   auto &ctx = MED->getASTContext();
   auto *dc = MED->getDeclContext();
-  auto foundMacros = TypeChecker::lookupMacros(
-      MED->getDeclContext(), MED->getMacro(),
-      MED->getLoc(), MacroRole::Declaration);
+  // Non-call declaration macros cannot be overloaded.
+  MacroRoles viableRoles(MacroRole::Declaration);
+  // In a code block context, a macro expansion decl can also expand to an
+  // arbitrary code item.
+  if (dc->isLocalContext() ||
+      dc->getContextKind() == DeclContextKind::FileUnit) {
+    viableRoles |= MacroRole::Expression;
+    viableRoles |= MacroRole::CodeItem;
+  }
+  SmallVector<MacroDecl *, 4> foundMacros;
+  auto foundRoles = TypeChecker::lookupMacros(
+      MED->getDeclContext(), MED->getMacro(), MED->getLoc(), viableRoles,
+      foundMacros);
   if (foundMacros.empty()) {
     MED->diagnose(diag::macro_undefined, MED->getMacro().getBaseIdentifier())
         .highlight(MED->getMacroLoc().getSourceRange());
-    return {};
+    return nullptr;
   }
+
+  auto expandAsMacroExpansionExpr = [&]() -> BraceStmt * {
+    Expr *mee = MED->createExpr();
+    TypeChecker::typeCheckExpression(mee, dc, {},
+                                     TypeCheckExprFlags::IsExprStmt |
+                                         TypeCheckExprFlags::IsDiscarded);
+    ASTNode expansion = mee;
+    // If top-level, wrap it in a `TopLevelCodeDecl(BraceStmt(...))`.
+    auto innermostDC = dc;
+    if (dc->getContextKind() == DeclContextKind::FileUnit) {
+      auto *tlcd = new (ctx) TopLevelCodeDecl(
+          dc, BraceStmt::createImplicit(ctx, {expansion}));
+      tlcd->setDeclContext(dc);
+      expansion = tlcd;
+      innermostDC = tlcd;
+    }
+    auto codeBlockItemList = BraceStmt::createImplicit(ctx, {expansion});
+    MED->setRewritten(codeBlockItemList);
+    TypeChecker::postTypeCheckMacroExpansion(mee, innermostDC);
```

Comment:
It's interesting that this doesn't happen as part of type-checking the outer context. Is this because at some outer level (say, walking all of the code items in the enclosing `BraceStmt`) we should also be walking into macro expansions? Similar to my comment earlier about having a semantic query that flattens all of the macro expansions so you walk a semantic view vs. the syntactic view?

Reply:
The `BraceStmt` isn't the problem because it was only a means of code block storage in `MacroExpansionDecl`, and was never visited by any walker (the walker skips this level). In #63553 I placed `BraceStmt` with `ArrayRef<ASTNode>` and the need for `postTypeCheck...` still remains.

I think the reason is because those post-checking passes are all triggered from the AST verifier, but the verifier isn't triggering macro expansions.